[Doc] Cross language page (#10471)
ray-project/ray: doc/source/cross-language.rst

> Cross Language

Reviewer: ```suggestion
Cross-language programming
```

==============

> This page will cover some more cross language examples of using Ray's flexible programming model.

Reviewer: ```suggestion
This page will show you how to use Ray's cross-language programming feature.
```

Python calling Java
-------------------

> You can write a Java class as follows:

Reviewer: ```suggestion
Suppose we have a Java static method and a Java class as follows:
```

.. code-block:: java

  // A regular Java class.
  public class Counter {

    private int value = 0;

    public int increment() {
      this.value += 1;
      return this.value;
    }

>     public static int add(int a, int b) {

Reviewer: Maybe put this in a different class. E.g., `Math.add`. It's a bit weird to put it here.

Author: OK. I will add a new class for the `add` function.

      return a + b;
    }
  }

> Then you can create the Java actor from Python, or call Java remote function
> from Python:

Reviewer: ```suggestion
Then, in Python, we can call the above Java remote function, or create an actor from the above Java class.
```

.. code-block:: python

  import ray

  ray.init(_include_java=True, _load_code_from_local=True)

  # Define a Java class.
  counter_class = ray.java_actor_class(

>         "<your java package>.Counter")

Reviewer: we can add something like `package io.ray.demo` in the Java example, and use the real package here.

Author: Good suggestion!

  counter = counter_class.remote()
  obj_ref1 = counter.increment.remote()
  assert ray.get(obj_ref1) == 1
  obj_ref2 = counter.increment.remote()
  assert ray.get(obj_ref2) == 2

  # Define a Java function.
  add_function = ray.java_function(
        "<your java package>.Counter", "add")
  
  # Call the Java remote function.
  obj_ref3 = add_function.remote(1, 2)
  assert ray.get(obj_ref3) == 3

  ray.shutdown()
  
Java calling Python
-------------------

You can write a Python module as follows:

.. code-block:: python

  class Counter(object):
    def __init__(self):
        self.value = 0

    def increment(self):
        self.value += 1
        return self.value

  def add(a, b):
      return a + b

Then you can create Python actor from Java, or call Python remote function
from Java:

.. code-block:: java

  import io.ray.api.Ray;
  import io.ray.api.function.PyActorClass;
  import io.ray.api.function.PyActorMethod;
  import io.ray.api.function.PyFunction;
  import org.testng.Assert;

  public class MyRayApp {

    public static void main(String[] args) {
      Ray.init();
      
      // Define a Python class.
      PyActorClass actorClass = PyActorClass.of(

>           "<your python module>", "Counter");

Reviewer: also, use real python module here. so this example will be easier to understand.

Author: OK.

      Assert.assertEquals(objRef3.get(), 3);

      Ray.shutdown();
    }
  }

Cross-language data serialization
---------------------------------

The arguments and return values of ray call can be serialized & deserialized
automatically if their types are the following:
  
  - Primitive data types
      ===========   =======  =======
      MessagePack   Python   Java
      ===========   =======  =======
      nil           None     null
      bool          bool     Boolean
      int           int      Short / Integer / Long / BigInteger
      float         float    Float / Double
      str           str      String
      bin           bytes    byte[]
      ===========   =======  =======

  - Basic container types
      ===========   =======  =======
      MessagePack   Python   Java
      ===========   =======  =======
      array         list     Array
      ===========   =======  =======

  - Ray builtin types
      - ActorHandle

.. note::

  * Be aware of float / double precision between Python and Java. If Java use a
    float type to receive the input argument, the double precision Python data
    will be reduced to float precision in Java.

> You can write a Python function which returns the input data:

Reviewer: Add "The following example shows how to pass these types as parameters and how to return return these types."

Author: Thanks.

.. code-block:: python

  @ray.remote
  def py_return_input(v):
      return v

Then you can transfer the object from Java to Python, then returns from Python
to Java:

.. code-block:: java

  import io.ray.api.Ray;
  import io.ray.api.function.PyActorClass;
  import io.ray.api.function.PyActorMethod;
  import io.ray.api.function.PyFunction;
  import org.testng.Assert;

  public class MyRayApp {

    public static void main(String[] args) {
      Ray.init();

      Object[] inputs = new Object[]{
          true,  // Boolean
          Byte.MAX_VALUE,  // Byte
          Short.MAX_VALUE,  // Short
          Integer.MAX_VALUE,  // Integer
          Long.MAX_VALUE,  // Long

>           // BigInteger can support max value of 2^64-1, please refer to:

Reviewer: If there is such an limitation, maybe we can undocument `BigInteger` now. As it's not very useful.

Author: If the `BigInteger` value out of MessagePack's range, it will raise an exception. So, I just remove this comments?

Cross-language programming
==========================

This page will show you how to use Ray's cross-language programming feature.

Python calling Java
-------------------

Suppose we have a Java static method and a Java class as follows:

.. code-block:: java

  package io.ray.demo;

  public class Math {

    public static int add(int a, int b) {
      return a + b;
    }
  }

.. code-block:: java

  package io.ray.demo;

  // A regular Java class.
  public class Counter {

    private int value = 0;

    public int increment() {
      this.value += 1;
      return this.value;
    }
  }

Then, in Python, we can call the above Java remote function, or create an actor
from the above Java class.

.. code-block:: python

  import ray

>   ray.init(_include_java=True, _load_code_from_local=True)

Reviewer: `_include_java` will be removed soon, let's just remove it here as well

Author: Fixed.

  ray.shutdown()
  
Java calling Python
-------------------

Suppose we have a Python module as follows:

.. code-block:: python

  # ray_demo.py

  class Counter(object):
    def __init__(self):
        self.value = 0

    def increment(self):
        self.value += 1
        return self.value

  def add(a, b):
      return a + b

Then, in Java, we can call the above Python remote function, or create an actor
from the above Python class.

.. code-block:: java

  package io.ray.demo;

  import io.ray.api.ObjectRef;
  import io.ray.api.PyActorHandle;
  import io.ray.api.Ray;
  import io.ray.api.function.PyActorClass;
  import io.ray.api.function.PyActorMethod;
  import io.ray.api.function.PyFunction;
  import org.testng.Assert;

  public class JavaCallPythonDemo {

    public static void main(String[] args) {
      Ray.init();

      // Define a Python class.
      PyActorClass actorClass = PyActorClass.of(
          "ray_demo", "Counter");

      // Create a Python actor and call actor method.
      PyActorHandle actor = Ray.actor(actorClass).remote();
      ObjectRef objRef1 = actor.task(
          PyActorMethod.of("increment", int.class)).remote();

>       Assert.assertEquals(objRef1.get(), 1);

Reviewer: ```suggestion
      Assert.assertTrue(objRef1.get() == 1);
```
Unfortunately, `assertEquals` doesn't support comparing `Integer` with `int.

Author: This case use `int.class`, so the `objRef1.get()` will return an `int` not an `Integer`.

            o).remote();
        Assert.assertEquals(res.get(), o);
      }

      Ray.shutdown();
    }
  }

Cross-language exception stacks
-------------------------------

Suppose we have a Java package as follows:

.. code-block:: java

  package io.ray.demo;

  import io.ray.api.ObjectRef;
  import io.ray.api.Ray;
  import io.ray.api.function.PyFunction;

  public class MyRayClass {

    public static int raiseExceptionFromPython() {
      PyFunction<Integer> raiseException = PyFunction.of(
          "ray_exception", "raise_exception", Integer.class);
      ObjectRef<Integer> refObj = Ray.task(raiseException).remote();
      return refObj.get();
    }
  }

and a Python module as follows:

.. code-block:: python

  # ray_exception.py

  import ray

  @ray.remote
  def raise_exception():
      1 / 0

Then, run the following code:

.. code-block:: python

  # ray_exception_demo.py

  import ray

  ray.init(_include_java=True, _load_code_from_local=True)

  obj_ref = ray.java_function(
        "io.ray.demo.MyRayClass",
        "raiseExceptionFromPython").remote()
  ray.get(obj_ref)  # <-- raise exception from here.

  ray.shutdown()

> The exception will be:

Reviewer: ```suggestion
The exception stack will be:
```
